### PR TITLE
flash: sam0: Add support for SAME54

### PR DESCRIPTION
**Pull request by benpicco**
This adds support for the flash peripheral found in the SAME5x/SAMD5x
line of MCUs.
The peripheral is very similar to the one found on SAMD2x with only a few register names changed.

Lock regions are different for each sam0 sub-family, so move them to DTS.

split off from #14685

### CODE REVIEW

**Review by Sizurka — APPROVED** [review]

**Comment by jhedberg** [comment]
Any update on this PR? It seems to lack at least codeowner reviews, so @nashif @galak @nvlsianpu please have a look.

As a general comment, it seems like there's existing issues with `#if` vs `#ifdef` usage in this driver  which should be fixed. In fact, most of these can (and should) probably be updated to use `if (IS_ENABLED(CONFIG...)) {` to better catch compilation issues with all possible configurations.

File: drivers/flash/flash_sam0.c [hunk 1]
```diff
@@ -316,17 +343,26 @@ static int flash_sam0_write_protection(struct device *dev, bool enable)
 
 	flash_sam0_sem_take(dev);
 
-	for (offset = 0; offset < CONFIG_FLASH_SIZE * 1024;
-	     offset += LOCK_REGION_SIZE) {
-		*FLASH_MEM(offset) = 0U;
+	for (offset = 0; offset < CONFIG_FLASH_SIZE * 1024;  offset += LOCK_REGION_SIZE) {
```
Comment (nvlsianpu) [inline drivers/flash/flash_sam0.c]: I would revert previous line breaking

**Review by nvlsianpu — APPROVED** [review]
Just one nip, I think using #ifdef-ry here is OK, although `IS_ENABLED(CONFIG...)` is more powerful it is still fine as switching macros have logic states `exist or not exist`.

File: drivers/flash/flash_sam0.c (hunk 1 above)
Comment (benpicco) [reply drivers/flash/flash_sam0.c]: Alright!

**Pull request merged**